### PR TITLE
feat: wildcard pattern added for fields to find all non-null values

### PR DESCRIPTION
**Pull request by vyavdoshenko**
fixes: https://github.com/dragonflydb/dragonfly/issues/4937

### CODE REVIEW

**Comment by BagritsevichStepan** [comment]
Please do merge with master

File: src/core/search/base.h [hunk 1]
```diff
@@ -90,6 +90,11 @@ struct BaseIndex {
   // Returns true if the document was added / indexed
   virtual bool Add(DocId id, const DocumentAccessor& doc, std::string_view field) = 0;
   virtual void Remove(DocId id, const DocumentAccessor& doc, std::string_view field) = 0;
+
+  // Returns all documents matching the star query (all documents in the index)
+  virtual std::optional<std::vector<DocId>> GetAllResults() const {
```
Comment (BagritsevichStepan) [inline src/core/search/base.h:95]: Nit: we are not returning AllResults, we are returning NotNullable results, take in account that json can store nullable values like this `{"text":nulll}` and we will index this. So you need change a bit comment and name

File: src/core/search/ast_expr.h [hunk 2]
```diff
@@ -22,6 +22,9 @@ struct AstNode;
 // Matches all documents
 struct AstStarNode {};
 
+// Matches all fields in a document (like "*" in field context)
```
Comment (BagritsevichStepan) [inline src/core/search/ast_expr.h:25]: Nit: It doesn't match all fields — it matches all documents where this field has a non-null value

File: src/core/search/indices.h [hunk 3]
```diff
@@ -58,6 +62,21 @@ template <typename C> struct BaseStringIndex : public BaseIndex {
   // Returns all the terms that appear as keys in the reverse index.
   std::vector<std::string> GetTerms() const;
 
+  std::optional<std::vector<DocId>> GetAllResults() const override {
+    std::vector<DocId> result;
+    std::vector<DocId> temp;
```
Comment (BagritsevichStepan) [inline src/core/search/indices.h:67]: Why just not to use `flat_hash_set` for DocId? And at the end map it to the vector.

File: src/core/search/indices.h [hunk 4]
```diff
@@ -58,6 +62,21 @@ template <typename C> struct BaseStringIndex : public BaseIndex {
...
+    for (const auto& term : GetTerms()) {
+      if (auto* docs = Matching(term)) {
```
Comment (BagritsevichStepan) [inline src/core/search/indices.h:69]: Why we need to do this? Take in account that GetTerms returns vector of std::string, so it makes copy for each term. Also Matching has a constant.

We have an iterator for RaxTreeMap, so you can just traverse entries_, and merge all containers.

File: src/core/search/indices.h [hunk 5]
```diff
@@ -133,6 +152,32 @@ struct FlatVectorIndex : public BaseVectorIndex {
...
   const float* Get(DocId doc) const;
 
+  // Return all documents that have vectors in this index
+  std::optional<std::vector<DocId>> GetAllResults() const override {
+    std::vector<DocId> result;
+    size_t num_vectors = entries_.size() / dim_;
+    result.reserve(num_vectors);
+
+    for (DocId id = 0; id < num_vectors; ++id) {
+      // Check if the vector is not zero (all elements are 0)
```
Comment (BagritsevichStepan) [inline src/core/search/indices.h:162]: Not the best decision here — a valid vector can contain zero values. However, since we don't need to fix this for 1.29, let's merge it as is and address it later in a separate PR.

File: src/server/search/search_family_test.cc [hunk 6]
```diff
@@ -2343,4 +2343,304 @@ TEST_F(SearchFamilyTest, KnnSearchWithSortby) {
   EXPECT_THAT(resp, IsArray(2, "doc:3", "doc:1"));
 }
 
+TEST_F(SearchFamilyTest, SearchNonNullFields) {
```
Comment (BagritsevichStepan) [inline src/server/search/search_family_test.cc:2346]: Good work with writing a lot of tests. But let's add a test for json indexes where json has null values for TEXT, TAG, NUMERIC and for TEXT SORTABLE, TAG SORTABLE, NUMERIC SORTABLE and assert that for `@field:*` we will not return those documents

File: src/server/search/search_family_test.cc [hunk 7]
```diff
@@ -2343,4 +2343,304 @@ TEST_F(SearchFamilyTest, KnnSearchWithSortby) {
...
+  string vector2 = "\\x00\\x00\\x00\\x00\\x00\\x00\\x80\\x3f\\x00\\x00\\x00\\x00";  // [0,1,0]
+  string vector3 = "\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x80\\x3f";  // [0,0,1]
+
+  Run({"hset", "vec:1", "embedding", vector1});
+  Run({"hset", "vec:2", "embedding", vector2});
+  Run({"hset", "vec:3", "embedding", vector3});
+
+  // Testing star query with result limit
+  auto limit_result = Run({"ft.search", "text_idx", "*", "LIMIT", "0", "2"});
+  ASSERT_GE(limit_result.GetVec().size(), 5);                 // Total count + 2 docs with fields
```
Comment (BagritsevichStepan) [inline src/server/search/search_family_test.cc:2435]: I didn't get why do we need this all sorting stuff for this functionality but ok. Let's keep it

File: src/server/search/search_family_test.cc [hunk 8]
```diff
@@ -2343,4 +2343,304 @@ TEST_F(SearchFamilyTest, KnnSearchWithSortby) {
...
+  EXPECT_EQ(limit_result.GetVec()[0].GetInt(), 3);            // Total count is 3 (all matches)
+  EXPECT_EQ(limit_result.GetVec()[1].GetString(), "text:1");  // First doc
+  EXPECT_EQ(limit_result.GetVec()[3].GetString(), "text:2");  // Second doc
+
+  // Testing star query with sorting
+  auto price_desc_result = Run({"ft.search", "num_idx", "*", "SORTBY", "price", "DESC"});
+  ASSERT_EQ(price_desc_result.GetVec().size(), 7);
+  EXPECT_EQ(price_desc_result.GetVec()[1].GetString(), "num:3");  // Most expensive item first
```
Comment (BagritsevichStepan) [inline src/server/search/search_family_test.cc:2443]: Nit: If we have repeated parts in our asserts or expects, we usually implement a MATCHER. But let's not do that in this PR — we can refactor it later

File: src/core/search/search.cc [hunk 9]
```diff
@@ -304,6 +305,44 @@ struct BasicSearch {
...
+        if (auto* docs = text_index->Matching(term); docs) {
+          sub_results.push_back(IndexResult{docs});
+        }
+      }
+
+      return UnifyResults(std::move(sub_results), LogicOp::OR);
+    } else if (auto* tag_index = dynamic_cast<TagIndex*>(base_index)) {
+      vector<IndexResult> sub_results;
+      sub_results.reserve(tag_index->GetTerms().size());
+
```
[the remark this answers is not in the document]
Comment (vyavdoshenko) [reply src/core/search/search.cc]: Fixed.

File: src/server/search/search_family_test.cc [hunk 10]
```diff
@@ -2278,4 +2278,40 @@ TEST_F(SearchFamilyTest, PrefixSearchWithSynonyms) {
   EXPECT_THAT(resp, AreDocIds("doc:6"));  // Should only find macintosh
 }
 
+TEST_F(SearchFamilyTest, SearchNonNullFields) {
+  EXPECT_EQ(Run({"ft.create", "i1", "schema", "title", "text", "tags", "tag", "score", "numeric"}),
```
[the remark this answers is not in the document]
Comment (vyavdoshenko) [reply src/server/search/search_family_test.cc]: Fixed.

File: src/core/search/base.h (hunk 1 above)
Comment (vyavdoshenko) [reply src/core/search/base.h:95]: Fixed.

File: src/core/search/indices.h (hunk 3 above)
Comment (vyavdoshenko) [reply src/core/search/indices.h]: Fixed.

File: src/core/search/indices.h (hunk 4 above)
Comment (vyavdoshenko) [reply src/core/search/indices.h]: Fixed.

File: src/server/search/search_family_test.cc (hunk 6 above)
Comment (vyavdoshenko) [reply src/server/search/search_family_test.cc:2346]: Test was added

File: src/core/search/ast_expr.h (hunk 2 above)
Comment (vyavdoshenko) [reply src/core/search/ast_expr.h]: Fixed.

File: src/core/search/base.h (hunk 1 above)
Comment (vyavdoshenko) [reply src/core/search/base.h:95]: Fixed.

**Review by BagritsevichStepan — APPROVED** [review]
